src/Propulsion.MessageDb/ReaderCheckpoint.fs

```diff
@@ -0,0 +1,81 @@
+module Propulsion.MessageDb.ReaderCheckpoint
+
+open Npgsql
+open NpgsqlTypes
+open Propulsion.Feed
+open Propulsion.Infrastructure
+open System
+
+let createIfNotExists (conn : NpgsqlConnection, schema: string) =
+    let cmd = conn.CreateCommand()
+    cmd.CommandText <- $"
+      create table if not exists {schema}.propulsion_checkpoints (
+        stream_name text not null,
+        consumer_group text not null,
+        global_position bigint not null,
+        primary key (stream_name, consumer_group)
+      );
+    "
+    cmd.ExecuteNonQueryAsync() |> Async.AwaitTaskCorrect |> Async.Ignore<int>
+
+let commitPosition (conn : NpgsqlConnection, schema: string) (stream : string) (consumerGroup : string) (position : int64) = async {
+     let cmd = conn.CreateCommand()
+     cmd.CommandText <-
+         $"insert into {schema}.propulsion_checkpoints(stream_name, consumer_group, global_position)
+           values (@StreamName, @ConsumerGroup, @GlobalPosition)
+           on conflict (stream_name, consumer_group)
+           do update set global_position = @GlobalPosition;"
+     cmd.Parameters.AddWithValue("StreamName", NpgsqlDbType.Text, stream) |> ignore
+     cmd.Parameters.AddWithValue("ConsumerGroup", NpgsqlDbType.Text, consumerGroup) |> ignore
+     cmd.Parameters.AddWithValue("GlobalPosition", NpgsqlDbType.Bigint, position) |> ignore
+
+     let! ct = Async.CancellationToken
+     do! cmd.ExecuteNonQueryAsync(ct) |> Async.AwaitTaskCorrect |> Async.Ignore<int> }
+
+let tryGetPosition (conn : NpgsqlConnection, schema : string) (stream : string) (consumerGroup : string) = async {
+    let cmd = conn.CreateCommand()
+    cmd.CommandText <-
+        $"select global_position from {schema}.propulsion_checkpoints where stream_name = @StreamName and consumer_group = @ConsumerGroup"
+
+    cmd.Parameters.AddWithValue("StreamName", NpgsqlDbType.Text, stream) |> ignore
+    cmd.Parameters.AddWithValue("ConsumerGroup", NpgsqlDbType.Text, consumerGroup) |> ignore
+
+    use reader = cmd.ExecuteReader()
+    let! ct = Async.CancellationToken
+    let! hasRow = reader.ReadAsync(ct) |> Async.AwaitTaskCorrect
+    return if hasRow then Some (reader.GetInt64 0) else None }
+
+type NpgsqlCheckpointStore(connString : string, schema: string, consumerGroupName, defaultCheckpointFrequency) =
```

Comment:
In `SSS` this is simply named `Service`, happy to maintain that convention but thought this was clearer.

Reply:
Yes, but it should be applied also to the Cosmos and DDB ones if we're going to break that (I'm perfectly happy for it to happen)
I think the name `CheckpointStore` will be fine as this is a `type` nested in a `module` (in general for types I would agree with the `Npgsql` bit)

Reply:
(FYI there was originally no base interface and/or ability for the various checkpoint stores to be interoperable - the Cosmos impl was the only game in town)
The V2 -> V3 time is the perfect time to fix that naming